Library/Homebrew/cmd/upgrade.rb

```diff
@@ -41,7 +42,10 @@ def upgrade
       puts pinned.map { |f| "#{f.full_name} #{f.pkg_version}" } * ", "
     end
 
-    outdated.each { |f| upgrade_formula(f) }
+    outdated.each do |f|
+      upgrade_formula(f)
```

Comment:
We may need to tweak `upgrade_formula` to make it return success or failure. So we only cleanup formula which succeeds in upgrading.